controller/rancher/certificate_fetcher.go

```diff
@@ -0,0 +1,252 @@
+	if err != nil {
+		return nil, fmt.Errorf("Coudln't get certificate by name [%s]. Error: %#v", certName, err)
+	}
+	var cert client.Certificate
+	var certWithChain string
+	if len(certs.Data) >= 1 {
+		cert = certs.Data[0]
+		certWithChain = fmt.Sprintf("%s\n%s", cert.Cert, cert.CertChain)
+	}
+	return &config.Certificate{
+		Name: cert.Name,
+		Key:  cert.Key,
+		Cert: certWithChain,
+	}, nil
+}
+
+func (fetcher *RCertificateFetcher) UpdateEndpoints(lbSvc *metadata.Service, eps []client.PublicEndpoint) error {
+	opts := client.NewListOpts()
+	opts.Filters["uuid"] = lbSvc.UUID
+	opts.Filters["removed_null"] = "1"
+	lbs, err := fetcher.Client.LoadBalancerService.List(opts)
+	if err != nil {
+		return fmt.Errorf("Coudln't get LB service by uuid [%s]. Error: %#v", lbSvc.UUID, err)
+	}
+	if len(lbs.Data) == 0 {
+		logrus.Infof("Failed to find lb by uuid %s", lbSvc.UUID)
+		return nil
+	}
+	lb := lbs.Data[0]
+
+	toUpdate := make(map[string]interface{})
+	toUpdate["publicEndpoints"] = eps
+	logrus.Infof("Updating Rancher LB [%s] in stack [%s] with the new public endpoints [%v] ", lbSvc.Name, lbSvc.StackName, eps)
+	if _, err := fetcher.Client.LoadBalancerService.Update(&lb, toUpdate); err != nil {
+		return fmt.Errorf("Failed to update Rancher LB [%s] in stack [%s]. Error: %#v", lbSvc.Name, lbSvc.StackName, err)
+	}
+	return nil
+}
+
+func (fetcher *RCertificateFetcher) ReadAllCertificatesFromDir(certDir string) []*config.Certificate {
+	certs := []*config.Certificate{}
+
+	fetcher.mu.RLock()
+	for _, value := range fetcher.CertsCache {
+		certs = append(certs, value)
+	}
+	fetcher.mu.RUnlock()
+
+	return certs
+}
+
+func (fetcher *RCertificateFetcher) ReadDefaultCertificate(defaultCertDir string) *config.Certificate {
+	var currentDefCert *config.Certificate
+
+	fetcher.mu.RLock()
+	currentDefCert = fetcher.DefaultCert
+	fetcher.mu.RUnlock()
+
+	return currentDefCert
+}
+
+func (fetcher *RCertificateFetcher) LookForCertUpdates(certDir string, defaultCertDir string, doOnUpdate func(string)) {
+	fetcher.CertDir = certDir
+	fetcher.DefaultCertDir = defaultCertDir
+	lastUpdated := time.Now()
+	for {
+		logrus.Debugf("Start --- LookForCertUpdates polling dir %v", fetcher.CertDir)
+		forceUpdate := false
+		certsUpdatedFlag := false
+		logrus.Debugf("lastUpdated %v", lastUpdated)
+
+		if time.Since(lastUpdated).Seconds() >= fetcher.forceUpdateInterval {
+			logrus.Debugf("LookForCertUpdates: Executing force update as certs in cache have not been updated in: %v minutes", fetcher.forceUpdateInterval)
+			forceUpdate = true
+		}
+
+		//read the certs from the dir into tempMap
+		fetcher.tempCertsMap = make(map[string]*config.Certificate)
+		err := filepath.Walk(fetcher.CertDir, fetcher.readCertificate)
+		if err != nil {
+			logrus.Errorf("LookForCertUpdates: Error %v reading certs from cert dir  %v", err, fetcher.CertDir)
+		} else {
+			//compare with existing cache
+			if !reflect.DeepEqual(fetcher.CertsCache, fetcher.tempCertsMap) || forceUpdate {
```

Comment:
Nit: forceUpdate should be evaluated first as if it is true, there is no need to run the comparison.